### PR TITLE
feat (api): workflow notification

### PR DESCRIPTION
**Pull request by sguiheux**

### CODE REVIEW

File: engine/api/notification/user.go [hunk 1]
```diff
@@ -136,6 +136,144 @@ func GetUserEvents(db gorp.SqlExecutor, pb *sdk.PipelineBuild, previous *sdk.Pip
...
+					}
+				}
+
+				//Finally deduplicate everyone
+				removeDuplicates(&jn.Recipients)
+				events = append(events, getWorkflowEvent(jn, params))
+			case sdk.EmailUserNotification:
+				jn, ok := notif.Settings.(*sdk.JabberEmailUserNotificationSettings)
+				if !ok {
+					log.Error("notification.GetUserEvents[Email]> cannot deal with %s", notif)
```
Comment (bnjjj) [inline engine/api/notification/user.go]: missing continue ?

File: engine/api/notification/user.go [hunk 2]
```diff
@@ -136,6 +136,144 @@ func GetUserEvents(db gorp.SqlExecutor, pb *sdk.PipelineBuild, previous *sdk.Pip
...
+		params["cds.author"] = p
+	}
+
+	for _, notif := range wr.Workflow.Notifications {
+		if ShouldSendUserWorkflowNotification(notif, nr, previousWR) {
+			switch notif.Type {
+			case sdk.JabberUserNotification:
+				jn, ok := notif.Settings.(*sdk.JabberEmailUserNotificationSettings)
+				if !ok {
+					log.Error("notification.GetUserWorkflowEvents[Jabber]> cannot deal with %s", notif)
```
Comment (bnjjj) [inline engine/api/notification/user.go]: missing continue ?

File: engine/api/permission/permission.go [hunk 3]
```diff
@@ -118,6 +118,48 @@ func AccessToEnvironment(key, env string, u *sdk.User, access int) bool {
...
+	rows, err := db.Query(query, projectID, access)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return []sdk.User{}, nil
+		}
+		return []sdk.User{}, err
+	}
+	defer rows.Close()
+
+	users := []sdk.User{}
```
Comment (bnjjj) [inline engine/api/permission/permission.go]: just a detail but if you declare this on the top of this function you don't have to write `return []sdk.User{}, nil` just before. --> `return users, nil`

File: engine/api/workflow/dao_node_run.go [hunk 4]
```diff
@@ -425,6 +414,28 @@ func GetNodeRunBuildCommits(db gorp.SqlExecutor, store cache.Store, p *sdk.Proje
...
+	var nodeRun sdk.WorkflowNodeRun
+	var rr = NodeRun{}
+	if err := db.SelectOne(&rr, query, n.Name, workflowID, nr.VCSBranch, nr.Number, nr.WorkflowNodeID, nr.ID); err != nil {
+		return nodeRun, sdk.WrapError(err, "PreviousNodeRun> Cannot load previous RUN: %s [%s %d %s %d]", query, n.Name, workflowID, nr.VCSBranch, nr.Number, nr.WorkflowNodeID)
+	}
+	pNodeRun, errF := fromDBNodeRun(rr)
+	if errF != nil {
+		return nodeRun, sdk.WrapError(errF, "PreviousNodeRun> Cannot read node run")
+	}
+	nodeRun = *pNodeRun
```
Comment (bnjjj) [inline engine/api/workflow/dao_node_run.go]: sure that pNodeRun won't be nil event if there isn't error ?

Comment (sguiheux) [reply engine/api/workflow/dao_node_run.go]: yes

File: engine/api/notification/user.go [hunk 5]
```diff
@@ -136,6 +138,146 @@ func GetUserEvents(db gorp.SqlExecutor, pb *sdk.PipelineBuild, previous *sdk.Pip
...
+
+	//Compute notification
+	params := map[string]string{}
+	for _, p := range nr.BuildParameters {
+		params[p.Name] = p.Value
+	}
+	//Set PipelineBuild UI URL
+	params["cds.buildURL"] = fmt.Sprintf("%s/project/%s/workflow/%s/run/%d", uiURL, wr.Workflow.ProjectKey, wr.Workflow.Name, wr.Number)
+	if p, ok := params["cds.triggered_by.username"]; ok {
+		params["cds.author"] = p
```
Comment (yesnault) [inline engine/api/notification/user.go]: "cds.triggered_by.username"

**Pull request merged**